### PR TITLE
Made Mutect2 isActive much better for low allele fractions

### PR DESCRIPTION
**Pull request by davidbenjamin**
@takuto This is the change to improve high-depth sensitivity for mitochondria.

@meganshand @fleharty You are welcome to review as well if you'd like.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2Engine.java [hunk 1]
```diff
@@ -268,29 +269,54 @@ private static int getCurrentOrFollowingIndelLength(final PileupElement pe) {
...
+        final double digammaAlphaPlusBeta = Gamma.digamma(alpha + beta);
+        final double rho = digammaBeta - digammaAlphaPlusBeta;
+        final double tau = digammaAlpha - digammaAlphaPlusBeta;
+
+
+
+        final double betaEntropy = Beta.logBeta(alpha, beta) - (alpha - 1)*digammaAlpha - (beta-1)*digammaBeta + (alpha + beta - 2)*digammaAlphaPlusBeta;
+
+        // TODO: check if the stream is too expensive
+        final double result = -betaEntropy + rho * refCount + altQuals.stream().mapToDouble(qual -> {
```
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2Engine.java]: As I said above, I think we want to add entropies here

File: docs/mutect/mutect.tex [hunk 2]
```diff
@@ -261,26 +262,53 @@ \section{Finding Active Regions}
...
 
 \begin{equation}
-L({\rm no~variation}) = \prod_{i \in \mathcal{R}} (1 - \epsilon_i) \prod_{j \in \mathcal{A}} \epsilon_j 
+L({\rm no~variation}) = \prod_{i \in \mathcal{R}} (1 - \epsilon_i) \prod_{j \in \mathcal{A}} \epsilon_j \approx \prod_{j \in \mathcal{A}} \epsilon_j, 
 \end{equation}
+where the approximation amounts to ignoring the possibility that ref reads are actually alt, or, equivalently, giving each ref read infinite quality.  This is not necessary but it greatly speeds the computation implementation because, as we will see, we will only need to keep alt base qualities in memory.
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply docs/mutect/mutect.tex]: done

File: docs/mutect/mutect.tex [hunk 3]
```diff
@@ -261,26 +262,53 @@ \section{Finding Active Regions}
...
 \begin{equation}
 L(f) = \prod_{i \in \mathcal{R}} \left[ (1 -f)(1 - \epsilon_i) + f \epsilon_i \right] \prod_{j \in \mathcal{A}} \left[f(1 - \epsilon_j) + (1 - f) \epsilon_j \right]
+\approx (1-f)^{N_{\rm ref}} \prod_{j \in \mathcal{A}} \left[f(1 - \epsilon_j) + (1 - f) \epsilon_j \right],
 \end{equation}
-The terms that signify observed ref reads that are actually alt reads with an error and vice versa are negligible\footnote{We can set an upper bound on the error in the log likelihood by Taylor-expanding to first order.  The error turns out to be quite small.}  Then we get
+where we again assign infinite base quality to ref reads and have let $N_{\rm ref} = | \mathcal{R}|$.
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply docs/mutect/mutect.tex]: done

File: docs/mutect/mutect.tex [hunk 4]
```diff
@@ -261,26 +262,53 @@ \section{Finding Active Regions}
...
-\int_0^1 (1-f)^{|\mathcal{R}|}f^{|\mathcal{A}|} \, df = \frac{ \Gamma(|\mathcal{R}| + 1) \Gamma(|\mathcal{A}| + 1)}{\Gamma(|\mathcal{R}| + |\mathcal{A}| + 2)} = \frac{|\mathcal{R}|! |\mathcal{A}|!}{(|\mathcal{R}|+|\mathcal{A}|+1)!}
+q(f) \approx {\rm Beta}(f | \alpha, \beta), \quad \alpha = N_{\rm alt} + 1, \beta = N_{\rm ref} + 1.
 \end{equation}
-In the likelihood ratio the reference factors $\prod_{i \in \mathcal{R}} (1 - \epsilon_i)$ cancel, leaving a log-odds of
+Here the ``$+1$"s come from the pseudocounts, one ref and one alt, of a flat prior of $f$.  Then, following the usual recipe of averaging the log likelihood with respect to $f$ and re-exponentiating, we find that $z_n$ ``sees" the following distribution:
 \begin{align}
-{\rm LOD} \approx& \sum_{j \in \mathcal{A}} \left[ \log (1 - \epsilon_j) - \log \epsilon_j \right] + \log \frac{|\mathcal{R}|! |\mathcal{A}|!}{(|\mathcal{R}|+|\mathcal{A}|+1)!} \\
- \approx& -\sum_{j \in \mathcal{A}} \log \epsilon_j + \log \frac{|\mathcal{R}|! |\mathcal{A}|!}{(|\mathcal{R}|+|\mathcal{A}|+1)!},
+q(z_n) &\propto \left[ \epsilon_n \overline{\ln (1 - f)} \right]^{z_n} \left[ (1 - \epsilon_n) \overline{\ln f} \right]^{1 - z_n} \\
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply docs/mutect/mutect.tex]: You're right.

File: docs/mutect/mutect.tex [hunk 5]
```diff
@@ -261,26 +262,53 @@ \section{Finding Active Regions}
...
+&= \left[ \epsilon_n \rho \right]^{z_n} \left[ (1 - \epsilon_n) \tau \right]^{1 - z_n},
 \end{align}
-the first term of which is proportional to the sum of effective base qualities.
+where we have defined the standard Beta distribution moments (with respect to $q(f)$) $\rho \equiv \overline{\ln (1 - f)} = \psi(\beta) - \psi(\alpha + \beta)$ and $\tau \equiv \overline{\ln f} = \psi(\alpha) - \psi(\alpha + \beta)$. By inspection, we see that
+\begin{equation}
+q(z_n) = {\rm Bernoulli}(z_n | \gamma_n), \quad \gamma_n = \frac{ \rho \epsilon_n}{\rho \epsilon_n + \tau (1 - \epsilon_n)}.
+\end{equation}
+Then, Equation 10.3 of Bishop gives us the variational lower bound on $L(f)$:
+\begin{align}
+L(f) &\approx E_q \left[ \ln P({\rm reads}, f, \vz) \right] - {\rm entropy}[q(f)] - \sum_n {\rm entropy}[q(z_n)] \\
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply docs/mutect/mutect.tex]: Fixed

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2Engine.java (hunk 1 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2Engine.java]: done

**Comment by davidbenjamin** [comment]
Back to you @takutosato.  You caught a couple of whoppers.  Fortunately, the validations still look good after fixing them.

File: docs/mutect/mutect.tex [hunk 6]
```diff
@@ -281,27 +281,27 @@ \section{Finding Active Regions}
...
+where we have defined the standard Beta distribution moments (with respect to $q(f)$) $\ln \rho \equiv \overline{\ln (1 - f)} = \psi(\beta) - \psi(\alpha + \beta)$ and $\ln \tau \equiv \overline{\ln f} = \psi(\alpha) - \psi(\alpha + \beta)$. By inspection, we see that
 \begin{equation}
 q(z_n) = {\rm Bernoulli}(z_n | \gamma_n), \quad \gamma_n = \frac{ \rho \epsilon_n}{\rho \epsilon_n + \tau (1 - \epsilon_n)}.
 \end{equation}
 Then, Equation 10.3 of Bishop gives us the variational lower bound on $L(f)$:
 \begin{align}
-L(f) &\approx E_q \left[ \ln P({\rm reads}, f, \vz) \right] - {\rm entropy}[q(f)] - \sum_n {\rm entropy}[q(z_n)] \\
-&= -H(\alpha, \beta) + \rho N_{\rm ref} + \sum_n \left[ \gamma_n (\rho + \ln \epsilon_n) + (1 - \gamma_n)(\tau + \ln(1 - \epsilon_n) - H(\gamma_n) \right],
+L(f) &\approx E_q \left[ \ln P({\rm reads}, f, \vz) \right] + {\rm entropy}[q(f)] + \sum_n {\rm entropy}[q(z_n)] \\
+&= -H(\alpha, \beta) +  N_{\rm ref} \ln \rho + \sum_n \left[ \gamma_n \ln \left( \rho \epsilon_n \right) + (1 - \gamma_n) \ln \left( \tau (1 - \epsilon_n) \right) - H(\gamma_n) \right],
```
Comment (takutosato) [inline docs/mutect/mutect.tex]: entropies should have positive signs.

Comment (davidbenjamin) [reply docs/mutect/mutect.tex]: fixed

**Pull request merged**